lib/src/rules/implicit_reopen.dart

```diff
@@ -0,0 +1,231 @@
+    registry.addClassDeclaration(this, visitor);
+  }
+}
+
+class _Visitor extends SimpleAstVisitor {
+  final LintRule rule;
+
+  _Visitor(this.rule);
+
+  void reportLint(
+    NamedCompilationUnitMember member, {
+    required InterfaceElement target,
+    required InterfaceElement other,
+    required String reason,
+  }) {
+    rule.reportLintForToken(member.name,
+        arguments: [target.name, other.name, reason]);
+  }
+
+  @override
+  void visitClassDeclaration(ClassDeclaration node) {
+    var classElement = node.declaredElement;
+    if (classElement == null) return;
+    if (classElement.hasReopen) return;
+
+    var library = classElement.library;
+
+    // | subtype          |  supertype   | `extends`/`with` |
+    // | _none_ or `base` |  `interface` | lint             |
+    // | `base`           |  `final`     | lint             |
+    if (classElement.hasNoModifiers || classElement.isBase) {
+      var supertype = classElement.superElement;
+      if (supertype.library == library) {
+        if (supertype.isInterface || supertype.isInducedInterface) {
+          reportLint(node,
+              target: classElement, other: supertype!, reason: 'interface');
+          return;
+        }
+      }
+
+      for (var m in classElement.mixins) {
+        var mixin = m.element;
+        if (mixin.library != library) continue;
+        if (mixin.isInterface || mixin.isInducedInterface) {
+          reportLint(node,
+              target: classElement, other: mixin, reason: 'interface');
+          return;
+        }
+      }
+    }
+
+    if (classElement.isBase) {
+      var supertype = classElement.superElement;
+      if (supertype.library == library) {
+        if (supertype.isFinal || supertype.isInducedFinal) {
+          reportLint(node,
+              target: classElement, other: supertype!, reason: 'final');
+          return;
+        }
+      }
+
+      for (var m in classElement.mixins) {
+        var mixin = m.element;
+        if (mixin.library != library) continue;
+        if (mixin.isFinal || mixin.isInducedFinal) {
+          reportLint(node, target: classElement, other: mixin, reason: 'final');
+          return;
+        }
+      }
+    }
+  }
+}
+
+extension on InterfaceElement? {
+  List<InterfaceType> get interfaces => this?.interfaces ?? <InterfaceType>[];
+
+  bool get isFinal {
+    var self = this;
+    return self != null && self.isFinal;
+  }
+
+  /// A sealed declaration `D` is considered final if it has a direct `extends` or
+  /// `with` superinterface which is `final`, or it has a direct superinterface
+  /// which is `base` as well as a direct `extends` or `with` superinterface
+  /// which is `interface`.
+  bool get isInducedFinal {
+    if (!isSealed) return false;
+
+    if (superElement.isFinal) return true;
+    if (mixins.any((m) => m.element.isFinal)) return true;
+
+    for (var i in interfaces) {
+      if (i.element.isBase) {
+        if (superElement.isInterface) return true;
```

Comment:
I know you didn't ask for these... but little implementation nit with doing an extra `superelement + mixins.length` amount of work with every check if an interface element is base.

Reply:
Good catch!